### PR TITLE
Name Endpoints if api_name is None

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

As discussed, the api_name will be set to the name of the function. If the fn is None, set to "unnamed". If it is a lambda, it will be set to "lambda". Multiple lambdas will have an index appended to their name in the order that they're added (this logic already existed).

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/events.py [hunk 1]
```diff
@@ -224,6 +225,18 @@ def inner(*args, **kwargs):
             if isinstance(show_progress, bool):
                 show_progress = "full" if show_progress else "hidden"
 
+            if api_name is None:
+                if fn is not None:
```
Comment (abidlabs) [inline gradio/events.py:229]: There are a few edge cases where a user could supply a function, but it doesn't have a `__name__` (e.g. if a user supplies a `functools.partial`). We should also set the name to "unnamed" in that case

File: gradio/events.py [hunk 2]
```diff
@@ -224,6 +225,18 @@ def inner(*args, **kwargs):
...
+                    api_name = "".join(
+                        [
+                            s
+                            for s in fn.__name__
+                            if s not in set(string.punctuation) - {"-", "_"}
```
Comment (abidlabs) [inline gradio/events.py:234]: We could potentially use `utils.strip_invalid_filename_characters` for this

**Comment by abidlabs** [comment]
See small nit above, otherwise looks good! Thanks for taking care of this @freddyaboulton

File: gradio/events.py (hunk 1 above)
Comment (freddyaboulton) [reply gradio/events.py:229]: Yea or a class with a `__call__` method (reason why CI is failing)

**Comment by freddyaboulton** [comment]
Thanks for the comments @abidlabs ! lambda functions show up as "lambda", partial functions show up as "partial" and callable classes show up as the name of the class.

I'm setting `api_name=False` for some of the events that get auto-added to Interfaces beyond the predict function, e.g. hiding stop button, submit events on inputs that are submittable, etc

**Pull request merged**